packages/freezed/lib/src/freezed_generator.dart

```diff
@@ -247,27 +229,89 @@ Read here: https://github.com/rrousselGit/freezed/blob/master/packages/freezed/C
     return false;
   }
 
-  List<Parameter> _commonParametersBetweenAllConstructors(
+  List<Parameter> _commonPropertiesBetweenAllConstructors(
     List<ConstructorDetails> constructorsNeedsGeneration,
   ) {
     return constructorsNeedsGeneration.first.parameters.allParameters
         .map((parameter) {
-          var hasAnyFinalProperty = false;
-          for (final constructor in constructorsNeedsGeneration) {
-            final matchingParameter =
-                constructor.parameters.allParameters.firstWhereOrNull((p) {
-              return p.name == parameter.name && p.type == parameter.type;
-            });
+          final library = parameter.parameterElement!.library!;
+
+          String typeString(DartType type, {bool withNullability = true}) {
+            return resolveFullTypeStringFrom(
+              library,
+              type,
+              withNullability: withNullability,
+            );
+          }
+
+          var anyMatchingPropertyIsFinal = false;
+
+          var commonSupertypeDartType = parameter.parameterElement!.type;
+          var commonSubTypeDartType = parameter.parameterElement?.type;
+
+          String? commonSupertypeString;
+          String? commonSubtypeString;
+
+          for (final constructor in constructorsNeedsGeneration.skip(1)) {
+            final matchingParameter = constructor.parameters.allParameters
+                .firstWhereOrNull((p) => p.name == parameter.name);
 
             if (matchingParameter == null) return null;
-            if (matchingParameter.isFinal) hasAnyFinalProperty = true;
-          }
 
-          if (hasAnyFinalProperty) {
-            return parameter.copyWith(isFinal: true);
+            final matchingParameterType =
+                matchingParameter.parameterElement!.type;
+            if (matchingParameter.isFinal) anyMatchingPropertyIsFinal = true;
+
+            if (commonSupertypeDartType is FunctionType ||
+                commonSupertypeDartType.isDynamic) {
+              // If the type is a typedef, by finding the upper bound we would lose
+              // the initial definition. Therefore FunctionTypes are currently not
+              // supported for finding in finding common super types.
+              // => Resort back to type string matching.
+              commonSupertypeString ??= parameter.type!;
+              if (commonSupertypeString.contains('dynamic')) return null;
+
+              if (!typeStringsEqualIgnoringNullability(
+                  commonSupertypeString, matchingParameter.type!)) {
+                return null;
+              }
+
+              if (commonSupertypeDartType.isNullable !=
+                  matchingParameterType.isNullable) {
+                commonSupertypeString =
+                    typeStringWithNullability(commonSupertypeString);
+                commonSubtypeString =
+                    typeStringWithoutNullability(commonSupertypeString);
```

Comment:
should this be `typeStringWithoutNullability(commonSubtypeString);`

Reply:
I'll take a look at this again next week!